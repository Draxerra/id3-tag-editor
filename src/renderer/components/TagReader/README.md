# Tag Reader

## Functionality

The tag reader is a file upload component which parses the uploaded file using a ID3 parser.

### Props

`<tag-reader>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `tag-parser-fn` | (files: File[]) => object[] | A function to parse an array of files into an array of ID3 tags.
| `tags` | object[] | The array of parsed files. Can be bound to using `v-model`.

### Events

`<tag-reader>` emits the following events:

| event | args type | description
| --- | --- | ---
| `tags-updated` | object[] | The updated array of tag objects. Automatically emitted when using `v-model`.

### Slots

`<tag-reader>` exposes the followings slots:

| slot | props | description
| --- | --- | ---
| `default` | n/a | Default slot.

### Example

```html
<tag-reader v-model="tags"
            tag-parser-fn="tagParserFn" />
```

```ts
import Vue from 'vue';
import { Component } from "vue-property-decorator";

import { readTagsSync } from "taglib2";

import TagReader from '~/src/renderer/components/TagReader/TagReader.vue';

@Component({
    components: {
        TagReader,
    },
})
export default class GenericComponent extends Vue {
    private tags = [];

    private tagParserFn(files: File[]) {
        return files.map((file) => readTagsSync(file));
    }
}
```