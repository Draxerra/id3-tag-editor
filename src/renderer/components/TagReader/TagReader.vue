<template lang="pug">
    .TagReader(@dragover.prevent=""
               @drop.prevent="handleDrop")
        slot
</template>

<script lang="ts">
import Vue from "vue";
import { Component, Model, Prop } from "vue-property-decorator";

import { ITag } from "~/src/types/tag";

@Component
export default class TagReader extends Vue {
    @Prop({type: Function, required: true})
        private tagParserFn!: (files: File[]) => Promise<object[]> | object[];
    @Model("tags-updated", {type: Array, required: true}) private tags!: object[];

    private showFileDialog() {
        const input = document.createElement("input");
        input.type = "file";
        input.multiple = true;
        input.accept = "audio/*";
        input.click();
        input.addEventListener("change", (event) => {
            event.preventDefault();
            const target = event.target as HTMLInputElement;
            if (target.files && target.files.length) {
                this.getTags(Array.from(target.files));
            }
        });
    }

    private handleDrop(event: DragEvent) {
        const filteredFiles = Array.from(event.dataTransfer.files).filter((file) =>
            file.type.includes("audio"),
        );
        if (filteredFiles.length) {
            this.getTags(filteredFiles);
        } else {
            this.$emit("error", new Error("Invalid file type."));
        }
    }

    private async getTags(files: File[]) {
        try {
            const tags = await this.tagParserFn(files);
            if (tags.length) {
                this.$emit("tags-updated", tags);
            } else {
                this.$emit("error", new Error("No tags successfully parsed from files."));
            }
        } catch (e) {
            this.$emit("error", e);
        }
    }
}
</script>

<style lang="stylus" scoped>
.TagReader {
    height: 100%;
    width: 100%;
}
</style>
