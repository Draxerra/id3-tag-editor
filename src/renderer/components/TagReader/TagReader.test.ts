import { mount } from "@vue/test-utils";
import TagReader from "./TagReader.vue";

const tagParserFn = (files: File[]) => {
    return files.map((file) => ({
        name: file.name,
    }));
};

const asyncTagParserFn = (files: File[]) => {
    return Promise.resolve(files.map((file) => ({
        name: file.name,
    })));
};

const asyncTagParserFnError = () => Promise.reject(new Error("Parser Error"));

describe("Tag Reader Component", () => {

    it("is a Vue instance", () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn, tags: [] }});
        expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it("snapshot has not changed", () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn, tags: [] }});
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("emits tags-updated on valid drop with sync tag parser", async () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn, tags: [] }});

        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test_file", { type: "audio/mp3" })];

        wrapper.trigger("drop", {
            dataTransfer: {
                files,
            },
        });
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted("tags-updated")).toEqual([[[{ name: "test_file" }]]]);
    });

    it("emits tags-updated on valid drop with async tag parser", async () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn: asyncTagParserFn, tags: [] }});

        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test_file", { type: "audio/mp3" })];

        wrapper.trigger("drop", {
            dataTransfer: {
                files,
            },
        });
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted("tags-updated")).toEqual([[[{ name: "test_file" }]]]);
    });

    it("emits error on invalid file type with drop", () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn, tags: [] }});

        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test_file", { type: "text/csv" })];

        wrapper.trigger("drop", {
            dataTransfer: {
                files,
            },
        });
        expect(wrapper.emitted("error")).toEqual([[new Error("Invalid file type.")]]);
    });

    it("emits error on tagParserFn error with drop", async () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn: asyncTagParserFnError, tags: [] }});

        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test_file", { type: "audio/mp3" })];

        wrapper.trigger("drop", {
            dataTransfer: {
                files,
            },
        });
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted("error")).toEqual([[new Error("Parser Error")]]);
    });

    it("emits nothing on no files from tagParserFn", async () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn: () => [], tags: [] }});

        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test_file", { type: "audio/mp3" })];

        wrapper.trigger("drop", {
            dataTransfer: {
                files,
            },
        });
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted("error")).toEqual([[new Error("No tags successfully parsed from files.")]]);
    });

    it("emits tags-updated on valid file dialog", async () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn, tags: [] }});
        const buf = new ArrayBuffer(8);
        const files = [new File([buf], "test_file", { type: "audio/mp3" })];

        const eventListener = jest.fn();
        spyOn(document, "createElement").and.returnValue({
            addEventListener: eventListener,
            click: jest.fn(),
        });

        (wrapper.vm as any).showFileDialog();

        const change = eventListener.mock.calls.find((call) => call[0] === "change");
        if (change) {
            change[1]({
                preventDefault: jest.fn(),
                target: {
                    files,
                },
            });
            await wrapper.vm.$nextTick();
            expect(wrapper.emitted("tags-updated")).toEqual([[[{ name: "test_file" }]]]);
        }
    });

    it("emits nothing on no files with file dialog", async () => {
        const wrapper = mount(TagReader, { propsData: { tagParserFn, tags: [] }});

        const eventListener = jest.fn();
        spyOn(document, "createElement").and.returnValue({
            addEventListener: eventListener,
            click: jest.fn(),
        });

        (wrapper.vm as any).showFileDialog();

        const change = eventListener.mock.calls.find((call) => call[0] === "change");
        if (change) {
            change[1]({
                preventDefault: jest.fn(),
                target: {
                    files: [],
                },
            });
            await wrapper.vm.$nextTick();
            expect(wrapper.emitted("error")).toBeUndefined();
            expect(wrapper.emitted("tags-updated")).toBeUndefined();
        }
    });

});
