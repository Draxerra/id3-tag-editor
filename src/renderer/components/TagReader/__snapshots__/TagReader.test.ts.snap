// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Reader Component snapshot has not changed 1`] = `<div class="TagReader"></div>`;
